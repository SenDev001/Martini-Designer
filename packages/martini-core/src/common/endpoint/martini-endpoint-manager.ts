import { JsonRpcServer } from "@theia/core";
import { MartiniEvent } from "../event/martini-event";

export const martiniEndpointManagerPath = "/services/martini/endpoints";

export enum EndpointType {
    DIR_WATCHER = "dir-watcher",
    EMAIL = "email",
    FTP_CLIENT = "ftp-client",
    FTP_SERVER = "ftp-server",
    HTTP_FILTER = "http-filter",
    JABBER = "jabber",
    JMS_LISTENER = "jms-listener",
    KAFKA_LISTENER = "kafka-listener",
    MONGODB_LISTENER = "mongodb-listener",
    URL_ALIAS = "rest-alias",
    RSS = "rss",
    SCHEDULER = "scheduler",
    TRACKER_RESUBMIT = "tracker-resubmit",
    REDIS_LISTENER = "redis-listener"
}

export interface MartiniEndpoint {
    type: EndpointType;
    name: string;
    service: string;
    packageName: string;
    enabled: boolean;
    modifiable: boolean;
    documentType: string;
    track: boolean;
    replicated: boolean;
    status: "STARTED" | "STOPPED";
}

export namespace MartiniEndpoint {
    export function is(object: any): object is MartiniEndpoint {
        return (
            !!object &&
            typeof object === "object" &&
            "type" in object &&
            "name" in object &&
            "service" in object &&
            "packageName" in object &&
            "enabled" in object &&
            "modifiable" in object &&
            "status" in object
        );
    }
    export function getType(type: string): EndpointType | undefined {
        const typeKey: string | undefined = Object.keys(EndpointType).find(
            (key: string) => EndpointType[key as keyof typeof EndpointType] === type
        );
        return typeKey ? EndpointType[typeKey as keyof typeof EndpointType] : undefined;
    }
}

export const MartiniEndpointManager = Symbol("MartiniEndpointManager");

export interface MartiniEndpointManager extends JsonRpcServer<MartiniEndpointManagerClient> {
    getAll(packageName: string): Promise<MartiniEndpoint[]>;

    get(packageName: string, endpointName: string): Promise<MartiniEndpoint | undefined>;

    start(packageName: string, endpointName: string): Promise<void>;

    stop(packageName: string, endpointName: string): Promise<void>;

    delete(packageName: string, endpointName: string): Promise<void>;

    save(packageName: string, endpoint: MartiniEndpoint): Promise<void>;

    test(packageName: string, endpoint: MartiniEndpoint): Promise<void>;
}

export interface EndpointEvent extends MartiniEvent {
    name: string;
    packageName: string;
    event: "STARTED" | "STOPPED" | "DELETED" | "SAVED";
}

export namespace EndpointEvent {
    export function is(event: MartiniEvent): event is EndpointEvent {
        return event.type === "ENDPOINT";
    }
}

export interface MartiniEndpointManagerClient {
    onEvent(event: EndpointEvent): void;
}

export const getDisplayName = (type: EndpointType): string => {
    switch (type) {
        case EndpointType.DIR_WATCHER:
            return "Directory Watcher";
        case EndpointType.EMAIL:
            return "Email";
        case EndpointType.KAFKA_LISTENER:
            return "Kafka Listener";
        case EndpointType.SCHEDULER:
            return "Scheduler";
        case EndpointType.TRACKER_RESUBMIT:
            return "Tracker Resubmit";
        case EndpointType.JABBER:
            return "Jabber";
        case EndpointType.FTP_CLIENT:
            return "FTP Client";
        case EndpointType.FTP_SERVER:
            return "FTP Server";
        case EndpointType.JMS_LISTENER:
            return "JMS Listener";
        case EndpointType.RSS:
            return "RSS";
        case EndpointType.URL_ALIAS:
            return "URL Alias";
        case EndpointType.MONGODB_LISTENER:
            return "MongoDB Listener";
        case EndpointType.REDIS_LISTENER:
            return "Redis Pub/Sub Listener";
        case EndpointType.HTTP_FILTER:
            return "HTTP Filter";
        default:
            return "Unknown";
    }
};

export interface RssEndpoint extends MartiniEndpoint {
    schedule: string;
    onlyNew: boolean;
    rssUrl: string;
}

export interface EmailSettings {
    host: string;
    port: number;
    username: string;
    password: string;
    ssl: boolean;
}

export interface ReplyEmailSettings extends EmailSettings {
    from: string;
}

export interface EmailEndpoint extends MartiniEndpoint, EmailSettings {
    schedule: string;
    replicated: boolean;
    track : boolean;
    protocol: "imap" | "pop3";
    deleteOnReceive: boolean;
    sendOutputAsReply: boolean;
    sendReplyOnError: boolean;
    replyEmailSettings: ReplyEmailSettings;
}
