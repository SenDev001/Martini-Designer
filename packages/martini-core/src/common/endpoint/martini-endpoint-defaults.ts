import { EndpointType, MartiniEndpoint, RssEndpoint } from "./martini-endpoint-manager";

export const createDefaultEndpoint = (packageName: string, name: string, type: EndpointType): MartiniEndpoint => {
    switch (type) {
        case EndpointType.RSS:
            return createDefaultRssEndpoint(packageName, name);
        default:
            return {
                name,
                type,
                packageName,
                status: "STOPPED",
                enabled: true
            } as MartiniEndpoint;
    }
};

const createDefaultRssEndpoint = (packageName: string, name: string): RssEndpoint => ({
    name,
    type: EndpointType.RSS,
    packageName,
    service: "",
    modifiable: true,
    replicated: true,
    status: "STOPPED",
    documentType: "RSS",
    enabled: true,
    track: false,
    schedule: "repeating:1000",
    rssUrl: "",
    onlyNew: false
});
