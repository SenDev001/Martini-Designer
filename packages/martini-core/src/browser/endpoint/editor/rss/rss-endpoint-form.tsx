import { Field, Form, Formik, FormikErrors, useFormikContext } from "formik";
import { noop } from "lodash";
import messages from "martini-messages/lib/messages";
import * as React from "react";
import Collapsible from "react-collapsible";
import * as Yup from "yup";
import { MartiniEndpoint, RssEndpoint } from "../../../../common/endpoint/martini-endpoint-manager";
import { DocumentType } from "../../../../common/tracker/document-type-manager";
import { FormRow, OnFormChange, validateSchema } from "../../../form/form";
import "../../../yup/yup-ext";
import { GeneralConfigurationSection, generalSchema } from "../general-config-section";

export interface RssEndpointFormProps {
    endpoint: MartiniEndpoint;
    reset: boolean;
    documentTypeProvider: () => Promise<DocumentType[]>;
    onChange: (connection: MartiniEndpoint) => void;
    onValidate: (errors: FormikErrors<MartiniEndpoint>) => void;
}

export const RssEndpointForm: React.FC<RssEndpointFormProps> = ({
    endpoint,
    reset,
    documentTypeProvider,
    onChange,
    onValidate,
}) => {
    const schema = React.useMemo(() => Yup.object().shape<Partial<RssEndpoint>>({
        rssUrl: Yup.string()
            .required()
            .url(),
        schedule: Yup.string()
            .required()
            .test(
                "is-positive-integer",
                messages.must_be_greater_or_equal(1),
                value => {
                    if (value) {
                        const index = value.lastIndexOf(":");
                        if (index > -1) {
                            const sub = value.substring(index + 1, value.length + 1);
                            const interval = Number.parseInt(sub);
                            if (!isNaN(interval) && interval > 0)
                                return true;
                        }
                    }

                    return false;
                })
    }).concat(generalSchema), []);

    const validate = async (endpoint: RssEndpoint) => validateSchema(schema, endpoint, onValidate);

    return <Formik
        initialValues={endpoint}
        initialTouched={{
            rssUrl: true,
            schedule: true
        }}
        enableReinitialize={reset}
        validate={validate}
        validateOnMount={true}
        onSubmit={noop}
    >
        <Form>
            <OnFormChange onChange={onChange} />
            <GeneralConfigurationSection documentTypeProvider={documentTypeProvider} />
            <RssConfigurationSection />
        </Form>
    </Formik>;
};

const RssConfigurationSection: React.FC = () => {
    const [opened, setOpened] = React.useState(true);
    const handleFocus = () => setOpened(true);
    const context = useFormikContext<RssEndpoint>();
    const handleScheduleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        const interval = Number.parseInt(e.currentTarget.value);
        if (isNaN(interval))
            context.setFieldValue("schedule", `repeating:${e.currentTarget.value}`);
        else
            context.setFieldValue("schedule", `repeating:${interval * 1000}`);
    };
    const getInterval = (schedule: string) => {
        const index = schedule.lastIndexOf(":");
        if (index > -1) {
            const sub = schedule.substring(index + 1, schedule.length + 1);
            const interval = Number.parseInt(sub);
            if (!isNaN(interval))
                return (interval / 1000).toString();
            return sub;
        }
        return "0";
    };
    return <Collapsible
        trigger={messages.configuration_title("RSS")}
        open={opened}
        onClosing={() => setOpened(false)}
        onOpening={() => setOpened(true)}
    >
        <div
            style={{
                display: "grid",
                gridColumnGap: 10,
                gridTemplateColumns: "max-content 1fr",
            }}
            onFocus={handleFocus}
        >
            <FormRow name="rssUrl" label="URL" tooltip={messages.rss_url_tooltip}>
                <Field name="rssUrl" type="text" className="theia-input" />
            </FormRow>
            <FormRow
                name="schedule"
                label={messages.rss_polling_interval_field}
                tooltip={messages.rss_polling_interval_tooltip}
            >
                <Field
                    name="schedule"
                    type="number"
                    className="theia-input"
                    min="1"
                    value={getInterval(context.values.schedule)}
                    onChange={handleScheduleChange}
                />
            </FormRow>
            <FormRow
                name="onlyNew"
                label={messages.rss_only_new_field}
                tooltip={messages.rss_only_new_tooltip}
            >
                <Field name="onlyNew" type="checkbox" />
            </FormRow>
        </div>
    </Collapsible>;
};
