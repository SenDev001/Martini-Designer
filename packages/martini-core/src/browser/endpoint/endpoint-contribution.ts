import {
    Command,
    CommandContribution,
    CommandRegistry,
    MenuContribution,
    MenuModelRegistry,
    SelectionService
} from "@theia/core";
import { ApplicationShell, CommonCommands } from "@theia/core/lib/browser";
import { inject, injectable } from "inversify";
import messages from "martini-messages/lib/messages";
import { MartiniEndpoint, MartiniEndpointManager } from "../../common/endpoint/martini-endpoint-manager";
import { ConfirmDialog, createListMessage } from "../dialogs/dialogs";
import { DeleteCommand } from "../martini-ide-contribution";
import { NAVIGATOR_VIEW_CONTEXT_MENU } from "../navigator/martini-navigator-view-contribution";
import { ProgressService } from "../progress/progress-service";
import { EndpointEditor } from "./editor/endpoint-editor";

export const ENDPOINT_CATEGORY = "Endpoint";

export const StartEndpointCommand: Command = {
    id: "martini.startEndpoint",
    label: messages.start,
    iconClass: "martini-icon martini-endpoint-start-icon",
    category: ENDPOINT_CATEGORY
};

export const StopEndpointCommand: Command = {
    id: "martini.stopEndpoint",
    label: messages.stop,
    iconClass: "martini-icon martini-endpoint-stop-icon",
    category: ENDPOINT_CATEGORY
};

@injectable()
export class EndpointCommandContribution implements CommandContribution {
    @inject(SelectionService)
    private readonly selectionService: SelectionService;
    @inject(MartiniEndpointManager)
    private readonly endpointManager: MartiniEndpointManager;
    @inject(ProgressService)
    private readonly progressService: ProgressService;
    @inject(ApplicationShell)
    private readonly shell: ApplicationShell;

    registerCommands(commands: CommandRegistry): void {
        commands.registerCommand(StartEndpointCommand, {
            execute: async () => {
                const selection = this.selectionService.selection;
                if (isEndpointSelection(selection)) {
                    const endpoints = selection.filter(e => e.status !== "STARTED");

                    const dlg = new ConfirmDialog({
                        title: endpoints.length > 1 ? messages.start_multi_endpoints : messages.start_single_endpoint,
                        msg:
                            endpoints.length <= 1
                                ? messages.start_single_endpoint_question(endpoints[0].name)
                                : createListMessage(
                                      messages.start_multi_endpoints_question,
                                      endpoints.map(e => e.name)
                                  ),
                        ok: messages.start
                    });

                    const confirmed = await dlg.open();
                    if (confirmed) {
                        await this.progressService.showProgress(messages.starting_endpoint, async progress => {
                            let count = 1;
                            for (const endpoint of endpoints) {
                                if (progress.isCancelled()) break;
                                progress.report({
                                    message: endpoint.name,
                                    work: { done: count++, total: endpoints.length }
                                });
                                await this.endpointManager.start(endpoint.packageName, endpoint.name);
                            }
                        });
                    }
                }
            },
            isEnabled: () => {
                const selection = this.selectionService.selection;
                return isEndpointSelection(selection) && selection.some(e => e.status === "STOPPED");
            },
            isVisible: () => {
                const selection = this.selectionService.selection;
                return isEndpointSelection(selection) && selection.some(e => e.status === "STOPPED");
            }
        });
        commands.registerCommand(StopEndpointCommand, {
            execute: async () => {
                const selection = this.selectionService.selection;
                if (isEndpointSelection(selection)) {
                    const endpoints = selection.filter(e => e.status !== "STOPPED");

                    const dlg = new ConfirmDialog({
                        title: endpoints.length > 1 ? messages.stop_multi_endpoints : messages.stop_single_endpoint,
                        msg:
                            endpoints.length <= 1
                                ? messages.stop_single_endpoint_question(endpoints[0].name)
                                : createListMessage(
                                      messages.stop_multi_endpoints_question,
                                      endpoints.map(e => e.name)
                                  ),
                        ok: messages.stop
                    });

                    const confirmed = await dlg.open();
                    if (confirmed) {
                        await this.progressService.showProgress(messages.stopping_endpoint, async progress => {
                            let count = 1;
                            for (const endpoint of endpoints) {
                                if (progress.isCancelled()) break;
                                progress.report({
                                    message: endpoint.name,
                                    work: { done: count++, total: endpoints.length }
                                });
                                await this.endpointManager.stop(endpoint.packageName, endpoint.name);
                            }
                        });
                    }
                }
            },
            isEnabled: () => {
                const selection = this.selectionService.selection;
                return isEndpointSelection(selection) && selection.some(e => e.status === "STARTED");
            },
            isVisible: () => {
                const selection = this.selectionService.selection;
                return isEndpointSelection(selection) && selection.some(e => e.status === "STARTED");
            }
        });
        commands.registerHandler(DeleteCommand.id, {
            execute: async () => {
                const selection = this.selectionService.selection;
                if (isEndpointSelection(selection)) {
                    const endpoints = selection.filter(e => e.packageName !== "core");

                    const dlg = new ConfirmDialog({
                        title: endpoints.length > 1 ? messages.delete_multi_endpoints : messages.delete_single_endpoint,
                        msg:
                            endpoints.length <= 1
                                ? messages.delete_single_endpoint_question(endpoints[0].name)
                                : createListMessage(
                                      messages.delete_multi_endpoints_question,
                                      endpoints.map(e => e.name)
                                  ),
                        ok: messages.delete
                    });

                    const confirmed = await dlg.open();
                    if (confirmed) {
                        await this.progressService.showProgress(messages.deleting_endpoint, async progress => {
                            let count = 1;
                            for (const endpoint of endpoints) {
                                if (progress.isCancelled()) break;
                                progress.report({
                                    message: endpoint.name,
                                    work: { done: count++, total: endpoints.length }
                                });
                                await this.endpointManager.delete(endpoint.packageName, endpoint.name);
                            }
                        });
                    }
                }
            },
            isEnabled: () => {
                const selection = this.selectionService.selection;
                return isEndpointSelection(selection) && selection.some(e => e.packageName !== "core");
            },
            isVisible: () => {
                const selection = this.selectionService.selection;
                return isEndpointSelection(selection) && selection.some(e => e.packageName !== "core");
            }
        });

        const findHandler = {
            execute: () => {
                if (this.shell.currentWidget instanceof EndpointEditor) this.shell.currentWidget.showSearchBox();
            },
            isEnabled: () => this.shell.currentWidget instanceof EndpointEditor
        };
        commands.registerHandler(CommonCommands.FIND.id, findHandler);
        commands.registerHandler("actions.find", findHandler);
    }
}

export const isEndpointSelection = (selection: any): selection is MartiniEndpoint[] => {
    return selection instanceof Array && selection.find(e => MartiniEndpoint.is(e));
};

@injectable()
export class EndpointMenuContribution implements MenuContribution {
    registerMenus(menus: MenuModelRegistry): void {
        menus.registerMenuAction(NAVIGATOR_VIEW_CONTEXT_MENU, {
            commandId: StartEndpointCommand.id
        });
        menus.registerMenuAction(NAVIGATOR_VIEW_CONTEXT_MENU, {
            commandId: StopEndpointCommand.id
        });
    }
}
