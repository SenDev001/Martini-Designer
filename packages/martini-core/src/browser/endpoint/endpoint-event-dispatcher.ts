import {Emitter, Event} from "@theia/core";
import {injectable} from "inversify";
import {EndpointEvent, MartiniEndpointManagerClient} from "../../common/endpoint/martini-endpoint-manager";

export const EndpointEventDispatcher = Symbol("EndpointEventDispatcher");

export interface EndpointEventDispatcher {
    onEndpointEvent: Event<EndpointEvent>;
}

@injectable()
export class EndpointEventDispatcherClient
    implements MartiniEndpointManagerClient, EndpointEventDispatcher {
    private readonly onEndpointEventEmitter = new Emitter<EndpointEvent>();
    readonly onEndpointEvent = this.onEndpointEventEmitter.event;

    onEvent(event: EndpointEvent): void {
        this.onEndpointEventEmitter.fire(event);
    }
}
