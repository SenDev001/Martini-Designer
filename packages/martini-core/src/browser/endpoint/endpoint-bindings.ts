import { CommandContribution, MenuContribution } from "@theia/core";
import { OpenHandler, WebSocketConnectionProvider, WidgetFactory } from "@theia/core/lib/browser";
import { interfaces } from "inversify";
import { MartiniEndpointManager, martiniEndpointManagerPath } from "../../common/endpoint/martini-endpoint-manager";
import {
    NavigatorContentProviderContribution,
    NavigatorLabelProviderContribution,
    NavigatorOpenHandler
} from "../navigator/martini-navigator-view-widget";
import { WizardContribution } from "../wizard/wizard-contribution";
import {
    EndpointEditor,
    EndpointEditorNavigatorOpenHandler,
    EndpointEditorOpenHandler,
    EndpointEditorOptions
} from "./editor/endpoint-editor";
import { EndpointCommandContribution, EndpointMenuContribution } from "./endpoint-contribution";
import { EndpointEventDispatcher, EndpointEventDispatcherClient } from "./endpoint-event-dispatcher";
import { EndpointNavigatorContentProvider, EndpointNavigatorLabelProvider } from "./endpoint-navigator-contribution";
import { EndpointWizard, EndpointWizardContribution, EndpointWizardPage } from "./wizard/endpoint-wizard";

export const bindEndpointBindings = (bind: interfaces.Bind) => {
    bind(EndpointEventDispatcherClient)
        .toSelf()
        .inSingletonScope();
    bind(EndpointEventDispatcher).toService(EndpointEventDispatcherClient);
    bind(MartiniEndpointManager)
        .toDynamicValue(ctx => {
            const connection = ctx.container.get(WebSocketConnectionProvider);
            return connection.createProxy(martiniEndpointManagerPath, ctx.container.get(EndpointEventDispatcherClient));
        })
        .inSingletonScope();
    bind(CommandContribution).to(EndpointCommandContribution);
    bind(MenuContribution).to(EndpointMenuContribution);
    bind(NavigatorContentProviderContribution).to(EndpointNavigatorContentProvider);
    bind(NavigatorLabelProviderContribution).to(EndpointNavigatorLabelProvider);

    bind(WizardContribution)
        .to(EndpointWizardContribution)
        .inSingletonScope();
    bind(EndpointWizard).toSelf();
    bind(EndpointWizardPage).toSelf();
    bind("Factory<EndpointWizard>").toAutoFactory(EndpointWizard);

    bind(WidgetFactory)
        .toDynamicValue(({ container }) => ({
            id: EndpointEditor.FACTORY_ID,
            createWidget: (options: EndpointEditorOptions) => {
                const child = container.createChild();
                child.bind(EndpointEditorOptions).toConstantValue(options);
                child.bind(EndpointEditor).toSelf();
                return child.get(EndpointEditor);
            }
        }))
        .inSingletonScope();

    bind(OpenHandler)
        .to(EndpointEditorOpenHandler)
        .inSingletonScope();
    bind(NavigatorOpenHandler)
        .to(EndpointEditorNavigatorOpenHandler)
        .inSingletonScope();
};
