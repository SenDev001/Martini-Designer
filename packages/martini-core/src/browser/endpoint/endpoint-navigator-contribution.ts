import {inject, injectable, postConstruct} from "inversify";
import * as React from "react";
import {MartiniEndpoint, MartiniEndpointManager} from "../../common/endpoint/martini-endpoint-manager";
import {PartialMartiniPackage} from "../../common/package/martini-package-manager";
import {
    NavigatorContentProviderContribution,
    NavigatorLabelProviderContribution
} from "../navigator/martini-navigator-view-widget";
import {AbstractTreeContentProvider, AbstractTreeLabelProvider, OverlayIcon, UpdateDataEvent} from "../tree/base-tree";
import {EndpointEventDispatcher} from "./endpoint-event-dispatcher";

@injectable()
export class EndpointNavigatorContentProvider
    extends AbstractTreeContentProvider
    implements NavigatorContentProviderContribution {
    @inject(MartiniEndpointManager)
    private endpointManager: MartiniEndpointManager;
    @inject(EndpointEventDispatcher)
    private endpointEventDispatcher: EndpointEventDispatcher;
    priority = 1;

    @postConstruct()
    init() {
        this.toDispose.push(this.endpointEventDispatcher.onEndpointEvent(async e => {
            if (e.event !== "DELETED" && e.event !== "SAVED") {
                const updated = (await this.endpointManager.getAll(e.packageName)).find(
                    endpoint =>
                        endpoint.name === e.name && endpoint.packageName === e.packageName
                );

                if (updated) {
                    this.onDidUpdateEmitter.fire({
                        affects: element =>
                            MartiniEndpoint.is(element) &&
                            element.name === e.name &&
                            element.packageName === e.packageName,
                        update: _ => updated
                    } as UpdateDataEvent);
                }
            } else {
                this.onDidUpdateEmitter.fire(
                    (element: any) =>
                        PartialMartiniPackage.is(element) && element.name === e.packageName
                );
            }
        }));
    }

    canHandle(element: any): boolean {
        return (
            PartialMartiniPackage.is(element) || EndpointListTreeNode.is(element)
        );
    }

    async resolveChildren(parent: any): Promise<any[]> {
        if (PartialMartiniPackage.is(parent) && parent.status !== "UNLOADED")
            return [
                {
                    type: "endpoint-list-node",
                    martiniPackage: parent
                } as EndpointListTreeNode
            ];

        if (EndpointListTreeNode.is(parent)) {
            if (parent.martiniPackage.status === "UNLOADED") return [];
            const endpoints = await this.endpointManager.getAll(
                parent.martiniPackage.name
            );
            return endpoints;
        }

        return [];
    }

    hasChildren(parent: any): boolean {
        return EndpointListTreeNode.is(parent);
    }

    getParent(element: any) {
        if (EndpointListTreeNode.is(element)) return element.martiniPackage;
    }
}

export class EndpointNavigatorLabelProvider extends AbstractTreeLabelProvider
    implements NavigatorLabelProviderContribution {
    priority: 0;

    canHandle(element: any): boolean {
        return EndpointListTreeNode.is(element) || MartiniEndpoint.is(element);
    }

    getCaption(element: any): string | React.ReactNode[] | undefined {
        if (EndpointListTreeNode.is(element)) return "Endpoints";
        if (MartiniEndpoint.is(element)) return element.name;
    }

    getIconClass(element: any): string | undefined {
        if (EndpointListTreeNode.is(element))
            return `martini-tree-icon martini-endpoint-icon`;
        if (MartiniEndpoint.is(element))
            return `martini-tree-icon martini-${element.type}-endpoint-icon`;
    }

    getOverlayIcons(element: any): OverlayIcon[] | undefined {
        if (MartiniEndpoint.is(element)) {
            const endpoint = element;
            if (endpoint.status === "STARTED")
                return [
                    {
                        width: 7,
                        height: 8,
                        iconClass: "martini-started-overlay-icon",
                        position: "top-right"
                    }
                ];
            if (endpoint.status === "STOPPED")
                return [
                    {
                        width: 7,
                        height: 8,
                        iconClass: "martini-stopped-overlay-icon",
                        position: "top-right"
                    }
                ];
        }
    }
}

export interface EndpointListTreeNode {
    type: "endpoint-list-node";
    martiniPackage: PartialMartiniPackage;
}

export namespace EndpointListTreeNode {
    export function is(object: any): object is EndpointListTreeNode {
        return (
            !!object &&
            object.type === "endpoint-list-node" &&
            "martiniPackage" in object
        );
    }
}
