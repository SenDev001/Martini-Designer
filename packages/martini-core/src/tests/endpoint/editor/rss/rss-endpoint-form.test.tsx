require("reflect-metadata");
import * as Enzyme from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import { noop } from "lodash";
import * as React from "react";
import * as Renderer from "react-test-renderer";
import { RssEndpointForm } from "../../../../browser/endpoint/editor/rss/rss-endpoint-form";
import { createDefaultEndpoint } from "../../../../common/endpoint/martini-endpoint-defaults";
import { EndpointType, RssEndpoint } from "../../../../common/endpoint/martini-endpoint-manager";

Enzyme.configure({ adapter: new Adapter() });

const endpoint = createDefaultEndpoint("test", "test", EndpointType.RSS) as RssEndpoint;

test("RssEndpointForm should be rendered", () => {
    const component = Renderer.create(<RssEndpointForm
        endpoint={endpoint}
        documentTypeProvider={async () => []}
        onChange={noop}
        onValidate={noop}
        reset={false}
    />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
