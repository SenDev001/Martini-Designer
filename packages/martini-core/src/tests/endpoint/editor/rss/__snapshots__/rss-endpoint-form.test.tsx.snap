// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RssEndpointForm should be rendered 1`] = `
<form
  action="#"
  onReset={[Function]}
  onSubmit={[Function]}
>
  <div
    className="Collapsible"
  >
    <span
      className="Collapsible__trigger is-open"
      onClick={[Function]}
      onKeyPress={[Function]}
      style={null}
      tabIndex={null}
    >
      General Configuration
    </span>
    <div
      className="Collapsible__contentOuter"
      onTransitionEnd={[Function]}
      style={
        Object {
          "WebkitTransition": "none",
          "height": "auto",
          "msTransition": "none",
          "overflow": "hidden",
          "transition": "none",
        }
      }
    >
      <div
        className="Collapsible__contentInner"
      >
        <div
          onFocus={[Function]}
          style={
            Object {
              "display": "grid",
              "gridColumnGap": 10,
              "gridTemplateColumns": "max-content 1fr",
            }
          }
        >
          <label
            htmlFor="service"
            onContextMenu={[Function]}
            style={
              Object {
                "marginTop": "3px",
                "verticalAlign": "center",
              }
            }
            title="The service to execute when the endpoint is triggered."
          >
            Service
          </label>
          <div
            style={
              Object {
                "display": "grid",
                "gridTemplateColumns": "1fr max-content",
                "gridTemplateRows": "max-content 1fr",
              }
            }
          >
            <input
              className="theia-input"
              name="service"
              onBlur={[Function]}
              onChange={[Function]}
              readOnly={true}
              type="text"
              value=""
            />
            <input
              className="theia-button"
              type="button"
              value="Browse..."
            />
            <div
              className="form-error-message-wrapper"
            />
          </div>
          <label
            htmlFor="documentType"
            onContextMenu={[Function]}
            style={
              Object {
                "marginTop": "3px",
                "verticalAlign": "center",
              }
            }
            title="The document type to be registered to Tracker when this endpoint is triggered."
          >
            Document Type
          </label>
          <div
            style={
              Object {
                "display": "grid",
                "gridTemplateColumns": "1fr max-content",
                "gridTemplateRows": "max-content 1fr",
              }
            }
          >
            <input
              className="theia-input"
              name="documentType"
              onBlur={[Function]}
              onChange={[Function]}
              readOnly={true}
              type="text"
              value="RSS"
            />
            <input
              className="theia-button"
              onClick={[Function]}
              type="button"
              value="Browse..."
            />
            <div
              className="form-error-message-wrapper"
            />
          </div>
          <label
            htmlFor="enabled"
            onContextMenu={[Function]}
            style={
              Object {
                "marginTop": "3px",
                "verticalAlign": "center",
              }
            }
            title="Whether or not the endpoint should be started automatically when the Martini package is started."
          >
            Auto Start
          </label>
          <div
            style={
              Object {
                "display": "grid",
                "gridTemplateColumns": "1fr",
                "gridTemplateRows": "max-content 1fr",
              }
            }
          >
            <input
              checked={true}
              name="enabled"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value={true}
            />
            <div
              className="form-error-message-wrapper"
            />
          </div>
          <label
            htmlFor="track"
            onContextMenu={[Function]}
            style={
              Object {
                "marginTop": "3px",
                "verticalAlign": "center",
              }
            }
            title="Flag determining whether executions should be logged to Tracker."
          >
            Log to Tracker
          </label>
          <div
            style={
              Object {
                "display": "grid",
                "gridTemplateColumns": "1fr",
                "gridTemplateRows": "max-content 1fr",
              }
            }
          >
            <input
              checked={false}
              name="track"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value={false}
            />
            <div
              className="form-error-message-wrapper"
            />
          </div>
          <label
            htmlFor="replicated"
            onContextMenu={[Function]}
            style={
              Object {
                "marginTop": "3px",
                "verticalAlign": "center",
              }
            }
            title="If this endpoint is configured on a Martini instance that's running in a cluster, replicated will determine whether to run the endpoint on all instances, or only the elected leader node in the cluster. When this is checked, all instances will run the endpoint. When it's unchecked, only the leader node will run the endpoint."
          >
            Replicated
          </label>
          <div
            style={
              Object {
                "display": "grid",
                "gridTemplateColumns": "1fr",
                "gridTemplateRows": "max-content 1fr",
              }
            }
          >
            <input
              checked={true}
              name="replicated"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value={true}
            />
            <div
              className="form-error-message-wrapper"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="Collapsible"
  >
    <span
      className="Collapsible__trigger is-open"
      onClick={[Function]}
      onKeyPress={[Function]}
      style={null}
      tabIndex={null}
    >
      RSS Configuration
    </span>
    <div
      className="Collapsible__contentOuter"
      onTransitionEnd={[Function]}
      style={
        Object {
          "WebkitTransition": "none",
          "height": "auto",
          "msTransition": "none",
          "overflow": "hidden",
          "transition": "none",
        }
      }
    >
      <div
        className="Collapsible__contentInner"
      >
        <div
          onFocus={[Function]}
          style={
            Object {
              "display": "grid",
              "gridColumnGap": 10,
              "gridTemplateColumns": "max-content 1fr",
            }
          }
        >
          <label
            htmlFor="rssUrl"
            onContextMenu={[Function]}
            style={
              Object {
                "marginTop": "3px",
                "verticalAlign": "center",
              }
            }
            title="The URL hosting the RSS feed."
          >
            URL
          </label>
          <div
            style={
              Object {
                "display": "grid",
                "gridTemplateColumns": "1fr",
                "gridTemplateRows": "max-content 1fr",
              }
            }
          >
            <input
              className="theia-input"
              name="rssUrl"
              onBlur={[Function]}
              onChange={[Function]}
              type="text"
              value=""
            />
            <div
              className="form-error-message-wrapper"
            />
          </div>
          <label
            htmlFor="schedule"
            onContextMenu={[Function]}
            style={
              Object {
                "marginTop": "3px",
                "verticalAlign": "center",
              }
            }
            title="The interval in seconds, at which Martini checks for changes in the RSS feed."
          >
            Polling Interval (seconds)
          </label>
          <div
            style={
              Object {
                "display": "grid",
                "gridTemplateColumns": "1fr",
                "gridTemplateRows": "max-content 1fr",
              }
            }
          >
            <input
              className="theia-input"
              min="1"
              name="schedule"
              onBlur={[Function]}
              onChange={[Function]}
              type="number"
              value="1"
            />
            <div
              className="form-error-message-wrapper"
            />
          </div>
          <label
            htmlFor="onlyNew"
            onContextMenu={[Function]}
            style={
              Object {
                "marginTop": "3px",
                "verticalAlign": "center",
              }
            }
            title="When checked, only 'unread' entries in the RSS feed will be sent to the service."
          >
            Only New Entries
          </label>
          <div
            style={
              Object {
                "display": "grid",
                "gridTemplateColumns": "1fr",
                "gridTemplateRows": "max-content 1fr",
              }
            }
          >
            <input
              checked={false}
              name="onlyNew"
              onBlur={[Function]}
              onChange={[Function]}
              type="checkbox"
              value={false}
            />
            <div
              className="form-error-message-wrapper"
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</form>
`;
